### PR TITLE
[WIP] rebased atomic physics

### PR DESCRIPTION
**Pull request by BrianMarre**
modelling of atomic transitions and coupling to PIC-simulation in PIConGPU

current status: first implementation

Features necessary:
Basic Infrastructure:
- [x] atomic states in PIConGPU:
     - [x] use charge number to reduce number of atomic states
     - [ ] use current charge state number to reduce memory required/increase number of modeled levels [future Feature]
- [x] histogram of electron energy distribution in a supercell
    - [x] adaptive energy histogram
        - allows for gaps in histogram
        - variable bin widths according to relative error
        - parallel binning
- [x] Numerical derivative of arbitary order and arbitary sample points
     - [x] memory optimised reimplementation of Fornberg, 1988

Physics Implementation:
- [x] new stage for atomic population dynamics (done in conjunction with  sbastrakov)
    - [x] substage filling the histogram in parallel
    - [x] substage solving the rate equation parallel in all macro-ions
        - markov chain like monte carlo solver of rate equation
    - [x] energy conservation with feedback to pic-electron momentum
        - [x] resolving energy
        - [x] parallel in pic-electrons
    - [x] non constant electron energy distribution over time step[future feature]: requires reworking of histogram implementation
    - [x] collisional de-/excitation rate calculation in picongpu, based on FlyCHK
    - [ ] improved basic ionization
    - [X] autonoumous de-/exciation rate calculation, based on FlyCHK
    - [ ] auto-ionization, based on flyCHK
    - [ ] ionization crossections [furture feature]: due to time constraints
    - [ ] radiative ionisation and de-/exciation [future feature], physics questions remaining

Implementation steps:
- [x] implementation of basic infrastructure
- [x] core implementation of algorithm
- [x] basic example for usage of new particle attributes
- [x] implement population dynamics due to electrons
- [x] test case example
- [ ] show correctness by comparing to FlyCHK and SCFly
- [ ] implement population dynamics due to fields
- [ ] .... further components as time allows

### CODE REVIEW

**Comment by BrianMarre** [comment]
@psychocoderHPC could you disable the compile test for now?

**Comment by BrianMarre** [comment]
currently ongoing BugHunt results:
Bugs:
[ ] missing reset of currentBinWidth to global initial Value after moving boundaries in AdaptiveHistogram.hpp/getBinLeftBoundary()
[ ] relative error functor returns NaN due to yet unknown reason
Issues found:
[ ] rewrite of AdaptiveHistogram.hpp/getBinLeftBoundary() loop without negation, to improve clarity
[ ] rewrite of AdaptiveHistogram.hpp/getBinWidth "copy paste" loop

**Comment by sbastrakov** [comment]
@BrianMarre FYI NaNs can explain some of strange behavior you observed. E.g. all comparisons using < or == are false when one of the arguments is NaN, so e.g. with `x == NaN` you will have `x <= y` and `y <= x` both false at the same time.

**Comment by BrianMarre** [comment]
> @BrianMarre FYI NaNs can explain some of strange behavior you observed. E.g. all comparisons using < or == are false when one of the arguments is NaN, so e.g. with `x == NaN` you will have `x <= y` and `y <= x` both false at the same time.

That is good to know. Thanks, I was not sure what a NaN in a comparison would result in

**Comment by psychocoderHPC** [comment]
@BrianMarre Please provide the A100 template files for hemera in a separate PR.

File: share/picongpu/examples/atomicPhysics/include/picongpu/param/starter.param [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
+
+namespace picongpu
```
Comment (psychocoderHPC) [inline share/picongpu/examples/atomicPhysics/include/picongpu/param/starter.param:24]: In your example/test please only provide files with changes compared to the default param files.

**Comment by BrianMarre** [comment]
will close this PR for now, new PR ready for merging is #4134.

**Pull request closed without merging**